httpx/_client.py

```diff
@@ -81,31 +82,67 @@ def get_proxy_map(
     ) -> typing.Dict[str, Proxy]:
         if proxies is None:
             if trust_env:
-                return {
-                    key: Proxy(url=url)
-                    for key, url in get_environment_proxies().items()
-                }
+                env_proxies = typing.cast(
+                    typing.Dict[URLTypes, typing.Union[URLTypes, Proxy]],
+                    get_environment_proxies(),
+                )
```

Comment:
This casting is a bit confusing to me, without it I get this mypy error:

`Argument 1 to "_split_proxy_map" of "BaseClient" has incompatible type "Dict[str, str]"; expected "Dict[Union[URL, str], Union[Union[URL, str], Proxy]]"`

Which I would've thought it's not an error given that `str` is both `Union`s?

Reply:
I think it's due to the use of `Dict` - I bet that if we switch the annotation to use a `Mapping[Union[...], Union[...]]` then we won't have this error. This has got me a few times in the past already, and I'm sure we have more of these edge cases elsewhere… (But we're not seeing them because most of our tests aren't type checked)

Reply:
Thanks @florimondmanca, but it didn't do the trick on this particular instance. I guess it's fine to have that cast there, though a bit annoying.

Reply:
I think this is the same issue than the HeaderTypes PR - URLTypes is a union but mapping is invariant in key so dicts that don’t exactly have that union as keys don’t match. Meh!